### PR TITLE
Fix the setting of odata.type in the constructor

### PR DESCRIPTION
**Pull request by andrueastman**
## Summary
This PR closes https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/issues/242

This PR updates the SDK code generator to set the odata.type under the following conditions for type disambiguation.

1. The type derives from an abstract type
2. One of its base types is referenced as the type for a property in another type (except if the base is entity).
3. One of its base types is referenced as the type in an odata action in another type (except if the base is entity).

This change therefore makes the setting of the type [unnecessary](https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/blob/master/src/Microsoft.Graph/Generated/requests/MessageAttachmentsCollectionRequest.cs#L57) when we try to post to abstract collections. This has therefore been removed as well. 

## Background/Related issues
- https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/issues/36#issuecomment-529597352
- https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/312
- https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/586

## Generated code differences
V1 - https://github.com/microsoftgraph/msgraph-sdk-dotnet/pull/993
beta - https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/pull/296

## Related issues to be closed 

- [x] https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/issues/67
- [x] https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/issues/283
- [x] https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/issues/226
- [x] https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/909
- [x] https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/560
- [x] https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/598
- [x] https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/566

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/534)

### CODE REVIEW

File: Templates/templates/CSharp/Model/ComplexType.cs.tt [hunk 1]
```diff
@@ -82,9 +84,12 @@ if (attributeStringBuilder.Length > 0) {
...
     {
 <#
-// Generate a constructor to initialize the @odata.type property when this type is not abstract.
-if (!complex.IsAbstract)
-{
+        // Do not generate a cstor if this is abstract.
+        if (complex.IsAbstract)
+        {
+        }
+        else if (isBaseTypeReferenced)
```
Comment (baywet) [inline Templates/templates/CSharp/Model/ComplexType.cs.tt:91]:
```suggestion
        if (!complex.IsAbstract && isBaseTypeReferenced)
```
And remove the block above

File: Templates/templates/CSharp/Model/EntityType.cs.tt [hunk 2]
```diff
@@ -82,28 +85,28 @@ if (attributeStringBuilder.Length > 0) {
...
+        ///<summary>
+        /// The <#=entityName#> constructor
+        ///</summary>
         public <#=cstorTypeDeclaration#>()
         {
             this.ODataType = "<#=entity.FullName#>";
         }
-	<#
-		}
+    <#
```
Comment (baywet) [inline Templates/templates/CSharp/Model/EntityType.cs.tt:108]:
```suggestion
<#
```

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs [hunk 3]
```diff
@@ -337,6 +337,62 @@ public static bool HasEnumProperties(this OdcmClass complex)
...
+                return false;
+            }
+            else
+            {
+                var isReferencedInAction = odcmClass.Namespace.Classes
+                    .SelectMany(c => c.Methods)  
+                    .Where( m => !m.IsFunction) // only get the Actions
+                    .Any(m => m.Parameters
+                        .Any( param => param.Type.Name == odcmClass.Base.Name 
+                                  && param.Type.Name != "entity"));
```
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:360]:
```suggestion
                                  && "entity".Equals(param.Type.Name, StringComparison.OrdinalIgnoreCase)));
```
protects you from casing differences and the name being null

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs [hunk 4]
```diff
@@ -337,6 +337,62 @@ public static bool HasEnumProperties(this OdcmClass complex)
...
+
+                var isReferencedInClass = odcmClass.Namespace.Types
+                    .Select(someType => someType)
```
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:363]:
```suggestion
```

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs [hunk 5]
```diff
@@ -337,6 +337,62 @@ public static bool HasEnumProperties(this OdcmClass complex)
...
+                    .Where(someType => someType is OdcmClass)
```
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:364]:
```suggestion
                    .OfType<OdcmClass>()
```

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs [hunk 6]
```diff
@@ -337,6 +337,62 @@ public static bool HasEnumProperties(this OdcmClass complex)
...
+                    .Any(someType => ((OdcmClass) someType).Properties
```
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:365]:
```suggestion
                    .Any(someType => someType.Properties
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs (hunk 6 above)
Comment (andrueastman) [reply src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs]: Resolved via https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/534/commits/5078472debf1e32df30f74a5aa4db8ff0b8106a6

File: Templates/templates/CSharp/Model/EntityType.cs.tt (hunk 2 above)
Comment (andrueastman) [reply Templates/templates/CSharp/Model/EntityType.cs.tt]: Resolved via https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/534/commits/5078472debf1e32df30f74a5aa4db8ff0b8106a6

**Comment by andrueastman** [comment]
Thanks for the great feedback @baywet. I have updated the PR with the feedback changes.

**Review by baywet — APPROVED** [review]

**Pull request merged**